server/src/main/java/org/elasticsearch/cluster/coordination/LeaderChecker.java

```diff
@@ -0,0 +1,268 @@
+        } else if (lastPublishedDiscoveryNodes.nodeExists(request.getSender()) == false) {
+            logger.debug("leader check from unknown node: {}", request);
+            transportChannel.sendResponse(new CoordinationStateRejectedException("leader check from unknown node"));
+        } else {
+            logger.trace("handling {}", request);
+            transportChannel.sendResponse(Empty.INSTANCE);
+        }
+    }
+
+    private class CheckScheduler implements Releasable {
+
+        private final AtomicBoolean isClosed = new AtomicBoolean();
+        private final AtomicLong failureCountSinceLastSuccess = new AtomicLong();
+        private final DiscoveryNode leader;
+
+        CheckScheduler(final DiscoveryNode leader) {
+            this.leader = leader;
+        }
+
+        @Override
+        public void close() {
+            if (isClosed.compareAndSet(false, true) == false) {
+                logger.debug("already closed");
+            } else {
+                logger.debug("closed");
+            }
+        }
+
+        void handleWakeUp() {
+            if (isClosed.get()) {
+                logger.debug("closed check scheduler woken up, doing nothing");
+                return;
+            }
+
+            logger.trace("checking {} with [{}] = {}", leader, LEADER_CHECK_TIMEOUT_SETTING.getKey(), leaderCheckTimeout);
+
+            // TODO lag detection:
+            // In the PoC, the leader sent its current version to the follower in the response to a LeaderCheck, so the follower
+            // could detect if it was lagging. We'd prefer this to be implemented on the leader, so the response is just
+            // TransportResponse.Empty here.
+            transportService.sendRequest(leader, LEADER_CHECK_ACTION_NAME, new LeaderCheckRequest(transportService.getLocalNode()),
+                TransportRequestOptions.builder().withTimeout(leaderCheckTimeout).withType(Type.PING).build(),
+
+                new TransportResponseHandler<TransportResponse.Empty>() {
+                    @Override
+                    public void handleResponse(Empty response) {
+                        if (isClosed.get()) {
+                            logger.debug("closed check scheduler received a response, doing nothing");
+                            return;
+                        }
+
+                        failureCountSinceLastSuccess.set(0);
+                        scheduleNextWakeUp(); // logs trace message indicating success
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (isClosed.get()) {
+                            logger.debug("closed check scheduler received a response, doing nothing");
+                            return;
+                        }
+
+                        long failureCount = failureCountSinceLastSuccess.incrementAndGet();
+                        if (failureCount >= leaderCheckRetryCount) {
```

Comment:
should we treat `exp instanceof ConnectTransportException || exp.getCause() instanceof ConnectTransportException` specially as in NodesFaultDetection to fail quicker?

Reply:
Yes, good catch. I pushed 0e1eade1ca9 to do this.